### PR TITLE
Tegra warrior updates

### PR DESCRIPTION
**Pull request by dwalkes**
See details in [topmost commit](https://github.com/mendersoftware/meta-mender-community/pull/107/commits/ff4ebda38ce3337a5537d5624de780c911cfed4d) message

There are a lot of changes implemented with the warrior branch.  I thought about attempting to split this commit up and I could do this if desired.  The redundant boot components might be the most useful to split out.  I did include a [reference to the branch](https://github.com/Trellis-Logic/meta-mender-community/commits/warrior-before-rebase) which contains each individual commit.  Let me know if you'd like any reorganization done with this.

I've added a lot to the README.md which is warrior specific.  I think I'll need some subset of this duplicated on the mender hub page as well, or I could move any to a warrior specific section on the mender hub page.

Let me know if you have questions or requests for additional changes.

### CODE REVIEW

File: meta-mender-tegra/recipes-mender/tegra-state-scripts/files/redundant-boot-reboot-state-script [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+nv_update_engine --install 1>&2
+# Should never get here, nv_update_engine will reboot on its own after successful update
+rc=$?
+if [ ${rc} -ne 0 ]; then
+    echo "A failure occurred in nv_update_engine (rc ${rc}).. nv_update_engine did not reboot" >&2
+    rm /opt/ota_package/bl_update_payload
+    ln -s /opt/ota_package/bl_update_payload_current /opt/ota_package/bl_update_payload
+    exit 1
+fi
+exit 0
```
Comment (mirzak) [inline meta-mender-tegra/recipes-mender/tegra-state-scripts/files/redundant-boot-reboot-state-script]: Nice!

File: meta-mender-tegra/recipes-mender/tegra-state-scripts/files/LICENSE [hunk 2]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (mirzak) [inline meta-mender-tegra/recipes-mender/tegra-state-scripts/files/LICENSE]: Do not think it is necessary to provide your own copy of the Apache 2.0 License, and instead you can use the following in your recipe:

```
LIC_FILE_CHECKSUM = "${COMMON_LICENSE_DIR}/Apache-2.0;md5=89aea4e17d99a7cacdbeed46a0096b10"
```

File: meta-mender-tegra/README.md [hunk 3]
```diff
@@ -17,7 +17,7 @@ Download the source:
     $ repo init \
             -u https://github.com/mendersoftware/meta-mender-community \
             -m meta-mender-tegra/scripts/manifest-tegra.xml \
-            -b thud
+            -b warrior
```
Comment (mirzak) [inline meta-mender-tegra/README.md]: Can you also update the README structure to match the suggested structure that we discussed. You can use [this](https://github.com/mendersoftware/meta-mender-community/blob/warrior/meta-mender-raspberrypi/README.md) as a template

File: meta-mender-tegra/README.md [hunk 4]
```diff
@@ -28,3 +28,20 @@ Build
 
     $ bitbake core-image-base
 
+## Customizing Image Rootfs Size
+
+To customize the size reserved for your rootfs, add variable MENDER_STORAGE_TOTAL_SIZE_MB to your build/local.conf or image and define with a value which is twice the size of your desired rootfs in MB.
```
Comment (mirzak) [inline meta-mender-tegra/README.md]: I do not know if this is needed here, and should be covered in docs.mender.io pretty well.

File: meta-mender-tegra/README.md [hunk 5]
```diff
@@ -28,3 +28,20 @@ Build
...
+* The redundant boot components are upgraded through a mender reboot state script.  This means the upgrades are not performed during manual install.  To perform reboot state script updates using manual install, the mender-install-manual script is provided. Run this script with the same arguments you would provide to mender -install.  The system will include reboot state script content after performing the update.
+
+Both of these changes above require changes to the u-boot environment location, which are outlined in the mender hub conversation linked above and also 
+[this post](https://devtalk.nvidia.com/default/topic/1063652/jetson-tx2/mmcblk0boot1-usage-at-address-4177408-and-u-boot-parameter-storage-space-availability/) on the nvidia forum about my observation related to mmcblkboot1.  
+This means upgrade from sumo or thud would require changes to move u-boot environment location and most likely would also require redundant boot support implemented as a post update state script similar to what
+is done on the warrior branch but also including additional payload and tools from earlier branches to support redundant boot.  I have not actually attempted this yet myself.
```
Comment (mirzak) [inline meta-mender-tegra/README.md]: Nice information, and definitely think that this should also be part of the MH post. I would suggest to put it in the "Known Issues" section which is part of the template of MH posts.

File: meta-mender-tegra/recipes-core/images/tegra-minimal-initramfs.bbappend [hunk 6]
```diff
@@ -0,0 +1,5 @@
+# Prevents circular dependency on u-boot due to EXTRA_IMAGEDEPENDS.
+# See https://hub.mender.io/t/warrior-tegra-build-cicular-dependencies/757/5 and comments from Mirza
+EXTRA_IMAGEDEPENDS_remove = "u-boot"
+# Don't need mender to work with this image
+IMAGE_FSTYPES_remove = "mender mender.bmap"
```
Comment (mirzak) [inline meta-mender-tegra/recipes-core/images/tegra-minimal-initramfs.bbappend]: I think this specific problem has been fixed with this commit,

https://github.com/mendersoftware/meta-mender/commit/a944df86e0aa7e61385ea1867847f65b59f8aeb3

File: meta-mender-tegra/recipes-mender/tegra-state-scripts/files/redundant-boot-reboot-state-script [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+echo "Mounting new install partition ${new_boot_part}" >&2
+mkdir -p ${new_mender_boot_part_mountpoint}
+mount /dev/mmcblk0p${new_boot_part} ${new_mender_boot_part_mountpoint}
+mount -oremount,rw /
+cp ${new_mender_boot_part_mountpoint}/opt/ota_package/bl_update_payload_current /opt/ota_package/bl_update_payload_new
+rm /opt/ota_package/bl_update_payload
+ln -s /opt/ota_package/bl_update_payload_new /opt/ota_package/bl_update_payload
+nv_update_engine --install 1>&2
+# Should never get here, nv_update_engine will reboot on its own after successful update
+rc=$?
```
Comment (mirzak) [inline meta-mender-tegra/recipes-mender/tegra-state-scripts/files/redundant-boot-reboot-state-script]: If it it is never supposed to get there, I think one can drop the check of return code and simply fail by reaching the code. What do you think?

File: meta-mender-tegra/recipes-mender/tegra-state-scripts/files/LICENSE (hunk 2 above)
Comment (dwalkes) [reply meta-mender-tegra/recipes-mender/tegra-state-scripts/files/LICENSE]: OK agreed, will do, I was copying the example at https://github.com/mendersoftware/meta-mender/tree/master/meta-mender-demo/recipes-mender/example-state-scripts/files FYI.

File: meta-mender-tegra/README.md (hunk 3 above)
Comment (dwalkes) [reply meta-mender-tegra/README.md]: Yes, sorry I missed that.

File: meta-mender-tegra/README.md (hunk 4 above)
Comment (dwalkes) [reply meta-mender-tegra/README.md]: Good point.  It didn't work right/well before warrior on tegra.  Probably better in a mender hub known issues section.

File: meta-mender-tegra/recipes-mender/tegra-state-scripts/files/redundant-boot-reboot-state-script (hunk 7 above)
Comment (dwalkes) [reply meta-mender-tegra/recipes-mender/tegra-state-scripts/files/redundant-boot-reboot-state-script]: I agree, that's a better solution

File: meta-mender-tegra/README.md (hunk 4 above)
Comment (dwalkes) [reply meta-mender-tegra/README.md]: I've moved to the updated known issues section on [mender hub](https://hub.mender.io/t/nvidia-tegra-jetson-tx2/123)

File: meta-mender-tegra/README.md (hunk 5 above)
Comment (dwalkes) [reply meta-mender-tegra/README.md]: Done, also added a section on standalone updates and bootloader components to introduce the mender-install-manual script

File: meta-mender-tegra/recipes-core/images/tegra-minimal-initramfs.bbappend (hunk 6 above)
Comment (dwalkes) [reply meta-mender-tegra/recipes-core/images/tegra-minimal-initramfs.bbappend]: Agreed, removed

**Comment by dwalkes** [comment]
@mirzak please take another look, I think I've implemented all of the requested changes.  I've also updated the [mender hub page](https://hub.mender.io/t/nvidia-tegra-jetson-tx2/123)

**Pull request merged**